### ISSUE TITLE
Autocompletion is not Proposing Table Names after "from "

### ISSUE THREAD

**Issue by Lonniebiz** [dbeaver/dbeaver#7879]
**DBeaver Version:** 6.3.4.202002011957
**Database:** [Pervasive 13](https://docs.actian.com/psql/PSQLv13/)

In DBeaver 6.3.4.202002011957, and probably for a couple prior versions, the SQL editor has stopped proposing table-names as I type.

For example, if I type `select * from `, I want the very next letter I type to receive auto-completion proposals of table names that contain what I type after `from `, but it isn't.

Here are my global "SQL Completion" settings:
![DBeaver_Autocompletion](https://user-images.githubusercontent.com/151919/74144705-d5260a00-4bc2-11ea-89d2-c9ed3d091cca.png)

Please advise.

**Comment by uslss** [comment]
hi @Lonniebiz ,
what is your db?

**Comment by Lonniebiz** [comment]
@uslss Pervasive 13

**Comment by Lonniebiz** [comment]
Field-Name auto-completion is working great in Version 7.0.0.202003021717.

However, table-name auto-completion still isn't happening. For example, after typing `select * from x`, I'd expect to see a list of all table-names containing the letter `x`.

**Comment by LonwoLonwo** [comment]
@Lonniebiz still actual for you in new [DBeaver versions](https://dbeaver.io/download/)?

**Comment by Lonniebiz** [comment]
@LonwoLonwo Yes sir, I'm currently using Version 7.1.5.202008180601 via snap and I've never seen Dbeaver autocomplete a table-name after typing the word "from".

In my style of writing queries I always explicitly typing the table name (and a dot) before each (and every) field name I type. I do this because it prevents ambiguity when multiple tables (used within the same query) have the exact same field names. If you don't include the table name with the field, when there are multiple tables with the exact same field name within the query, the query will fail. So, I always include the table name. It also prevents that same ambiguity while I'm reading the query.

Obviously, any time someone has just got through typing the word "from", Dbeaver should anticipate that the next thing the user will type will be a table name. Yet, it never auto-completes this predictable need.

Additionally, Dbeaver always things I'm going to type a field name, when like I said very, I'm always typing the table name first, but perhaps that is a separate issue to the "from" isssue I'm reporting here.

**Comment by LonwoLonwo** [comment]
@Lonniebiz you can also try using ctrl+space in the place where you expect autocompletion. Or try to do this in another databases. 
![2020-09-04 10_19_30-Window](https://user-images.githubusercontent.com/45152336/92211544-4ac43100-ee99-11ea-9d9e-d98af8d120e9.png)
it work fine for me.
but most likely, unfortunately, due to the fact that the beaver does not support the Pervasive database in a special way, it is also difficult for him to give a list of tables

**Comment by Lonniebiz** [comment]
@LonwoLonwo 

When I type ctrl+space, instead of listing tables it only list the name of the whole database in the auto-completion list (even though I already have that same database selected on the Dbeaver tool bar).

Also, even when I type the databaseName followed by a dot, it still doesn't list the tables (not even if I type ctrl+space)

Note that Deaver has no issue listing all of the tables within the Pervasive database when navigating using Dbeaver's Database Navigator tool. It seems like if Dbeaver is capable of listing all those table names via the Database Navigator, it should also be able to accomplish that same task in auto-complete.

I understand how it can be difficult to fix an issue Dbeaver is having with a particular database if you yourself do not also have access to a Pervasive database. If you like, I can make a video showcasing this, and I am also available for online meetings.

**Comment by LonwoLonwo** [comment]
@Lonniebiz ok, let's try video. send it here

**Comment by uslss** [comment]
there is no update on issue for a long time. 
if issue is still actual for you - feel free to reopen it

**Comment by DustinMeldrumRoot** [comment]
Has anyone reached a conclusion on this issue? I'm a new user, curious if auto-prepending of the schema name and/or the ability to set and search a default schema for table name autocomplete is available.

**Comment by GertL** [comment]
I have the same issue with SyBase IQ. Is there a resolution?

**Comment by Lonniebiz** [comment]
So, today I made a video showcasing this issue:
https://youtu.be/6OW0liGSLTk

However, after reviewing that video, I realized what was happening. Basically, at the time that I was expecting "Auto-completion of Table Names", I was instead getting a list of **database names**. At that time, here's what my global auto-completion settings were:

![AutoCompleteSettings1](https://user-images.githubusercontent.com/151919/116898664-46a10c80-abfc-11eb-8126-106d12cbc15a.png)

After making the changes (highlighted in red, below), I was then able to achieve auto-completion of table-names by typing a portion of the table-name and then pressing ctrl+space_bar (as recommended by @LonwoLonwo):

![AutoCompleteSettings2](https://user-images.githubusercontent.com/151919/116899482-350c3480-abfd-11eb-8c76-794af6632d12.png)

Here's how I got to those global auto-completion settings above:
1) Right-click inside the SQL Editor.
2) Click "Preferences . . .".
3) In the left pane, nagivate to Editors > SQL Editors > Code Completion
4) Click  the "Global settings" hyperlink that's in the top-right corner.

**Comment by LonwoLonwo** [comment]
@Lonniebiz and it works with this settings?

**Comment by Lonniebiz** [comment]
After enabling "**Use short object names (omit schema/catalog**)", as described above, **automatic** auto-completion does NOT work. Meaning, it doesn't try to make suggestions as you're typing. 

However, if you hit ctrl+space_bar (after you've typed a few letters of the table name, it will then offer auto-completion suggestions that are indeed the table names in the database.

I wish there was a feature that I could set that would make the **ctrl+space_bar** step unnecessary.

Typical default auto-completion behavior would be for Dbeaver to just **automatically** offer suggestions as I'm typing the table name, and not require the **ctrl+space_bar** step.

However, thank you for telling me about **ctrl+space_bar**; at least it is a hotkey and not too terribly burdensome once you know it, but it would be more intuitive if it worked as I've described. The new user would benefit from the auto-completion without having to learn about **ctrl+space_bar** on github.

**Comment by GertL** [comment]
It fails when writing schema_name.Table_name. If I directly write table_name it is able to find the combination. 
e.g. writing CustomerPortfolio, it can find DW.CustomerPortfolio, but writing DW.CustomerPortfolio, it cannot find nor it has proposals. And that is also with the settings displayed above.

However, that is not very sustainable. :(

**Comment by Lonniebiz** [comment]
I noticed what @GertL is describing (I think). @GertL, what does it do if you type:
`select * from DW.DW.CustomerP` ?

For me, it would auto-complete at this point, producing:
`select * from DW.DW.CustomerPortfolio`

However, if I execute the query it fails due to "DW." being in there twice. If I take the extra "DW." out the query succeeds. Table name auto-completion will not occur without the extra erroneous "DW."

**Comment by GertL** [comment]
@Lonniebiz , `select * from DW.DW.CustomerP`  would have no proposal.

**Comment by Lonniebiz** [comment]
For me, given your example, mine would auto-completing "DW" twice before finally auto-completing table names. Like this:
`select * from DW.DW.tableName`

However, that query fails execution unless you change it to:
`select * from DW.tableName`

So, for me, the autocompletion of table-names only happens after the 2nd dot (which produces a query that fails execution).

**Comment by LonwoLonwo** [comment]
@Lonniebiz What structure has Pervasive? Connection-Database-Schema-Table? 

![2021-05-04 23_28_37-DBeaver Enterprise 21 0 4 - domain_1](https://user-images.githubusercontent.com/45152336/117065847-a567a280-ad30-11eb-8626-38720482f937.png)

Or Connection-Schema-Table? 
![2021-05-04 23_29_00-DBeaver Enterprise 21 0 4 - domain_1](https://user-images.githubusercontent.com/45152336/117065848-a6003900-ad30-11eb-9f9d-8709077faff9.png)

Is "MT" default schema in your example?

**Comment by Lonniebiz** [comment]
In Pervasive 13, there are two databases that I use. One is called MT and the other is called TESTING, which are highlighted in red below:

![psq](https://user-images.githubusercontent.com/151919/117081010-9bc84580-ad04-11eb-959a-6c10a6ce7635.png)

I use TESTING so rarely that I do not even have it registered in DBeaver.

In Dbeaver, when I add MT to the Database Navigator it looks like this:

![DatabaseNavigator](https://user-images.githubusercontent.com/151919/117081127-e053e100-ad04-11eb-8add-6cc697263ae5.png)

Its a bit confusing to me, because I think I named the connection MT too (which is the same name as the database I'm ultimately accessing). Both of these queries execute without issue:
`select * from customer`
`select * from mt.customer`

However, this query will **not** execute without error:
`select * from mt.mt.customer`

Yet, the auto-completion in Dbeaver thinks both of those MTs (mt.mt.tableName) are necessary before Dbeaver will offer table suggestions.

After I enabled "Use short object names (omit schema/catalog)", I'm now able to just type a portion of a table-name and then hit **ctrl+space_bar**, and Dbeaver will offer correct suggestions such as **mt.customers** (not mt.mt.customers), and after I select one of these suggestions it will simply input "customers" (alone; without "mt.") into my query (which is what I want, but wish I didn't have to hit **ctrl+space_bar** to achieve it). I wish it would offer those table names while I'm typing them instead.

My connection profile connects me directly to the MT database, so I don't understand why "mt.mt" is ever necessary. If I send pervasive a query with mt.mt.tableName in it, it doesn't expect but one of those "mt."s. Yet, the default auto-completion behavior in Dbeaver thinks both MTs (mt.mt.tableName) are necessary and will not offer table suggestions until both MTs are there.  I do not want my final query to have the database name anywhere in it. I only want the table name to be there. The name of the database I'm connecting to already implies which database the tables are in, so I want to avoid that unnecessary specification (that Pervasive doesn't require) in the SQL itself.

**Comment by LonwoLonwo** [comment]
How DBeaver works with autocomplete - you type database name with dot on the end (like "select * from MT.") and DBeaver propose you schema names.

Like we do it for PostgreSQL
![2021-05-10 14_39_55-](https://user-images.githubusercontent.com/45152336/117653914-b20b5100-b19d-11eb-96eb-f4abccf857ae.png)

Or DBeaver propose table names, if database doesn't have schemas, like in MySQL
![2021-05-10 14_42_13-](https://user-images.githubusercontent.com/45152336/117654185-14fce800-b19e-11eb-9487-6907a0654996.png)

so, in case structure database-schema-table, DBeaver will propose table names for schema name with dot on the end. But the problem is, as I can see now, that your database name equals schema name. DBeaver think, that "MT" is Database name. 
I guess, autocomplete should work for schema with not equal database name.

another question - does Pervasive have schemas in its structure?

DBeaver does not know, that propose in case no default schema.
![2021-05-10 14_50_27-DBeaver Enterprise 21 0 5 - _dvdrental_ Script-23](https://user-images.githubusercontent.com/45152336/117655100-4e822300-b19f-11eb-9178-0d51f5efd760.png)

And works for default schema:
![2021-05-10 14_32_29-DBeaver Enterprise 21 0 5 - _dvdrental_ Script-23](https://user-images.githubusercontent.com/45152336/117656079-92c1f300-b1a0-11eb-99e5-616c98c93231.png)

Do you know how to make schema default in Pervasive?

**Comment by Lonniebiz** [comment]
I've seen databases like IBM DB2, where each user has their own schema for what seems to be the exact same database. I guess they do that for authentication/permissions purposes.

I've never seen that type of thing in any version of Pervasive. In Pervasive 13, I see no mention of schema in its native tool's hierarchy:
![psql](https://user-images.githubusercontent.com/151919/117705359-913cfe80-b191-11eb-8a3c-72f60f8b0b3b.png)

In the Dbeaver connection profile, under Initialization, I do see a place to set default schema, but it is disabled (probably because the schema concept doesn't apply to Pervasive).

Right now, if I type "mt." nothing happens. Then if I hit **ctrl+space_bar**, it says "no proposals". However, if I do not include "mt." and just type the "first few letters of the table name" and then press  **ctrl+space_bar**, it makes proposals like "mt.tableName1", which upon selecting just enters in the table name alone (without "mt.").

In summary, I'd like it to just "propose table-names as I type them" without having to hit  **ctrl+space_bar**. However, you say the expected behavior is to only propose table names after the user has typed "mt." (for example). Neither my goal nor your expected behavior is happening.

**Comment by LonwoLonwo** [comment]
Hmm.
@Lonniebiz Can you try to omit schemas in driver settings and test again?

![2021-05-11 12_52_53-Window](https://user-images.githubusercontent.com/45152336/117796833-1985d700-b258-11eb-8cf4-6385cdb61be4.png)

![2021-05-11 12_53_30-Window](https://user-images.githubusercontent.com/45152336/117796835-1a1e6d80-b258-11eb-90f5-f891d6fdc760.png)

**Comment by Lonniebiz** [comment]
Wow! Thank you so much. After editing my connection settings to "Omit schema(s)", the auto-completion behavior is working exactly like I want. Now, by merely typing a small portion of a table-name (even if the characters I type are in the middle of the name) Dbeaver begins making correct proposals and it also narrows down those proposals the more I type. Perfect!

![Perfect](https://user-images.githubusercontent.com/151919/117816285-f9402300-b22b-11eb-86af-0b9470bdda2e.png)

Although this is the exact behavior **I** want, to be thorough, I must inform you that "mt." (throughout all our efforts) has never produced any auto-completion after that dot. And, even if you hit **ctrl+space_bar**, after typing "mt.cust" it would always just say "no proposals" (despite the confirmed existence of tables like "mt.customer").

I'm very satisfied, but if you decide to investigate this remaining issue further (for the purposes of increasing your wizard's ability to optimize Pervasive Database Connections by default), I will assist for the benefit of others.

**Comment by LonwoLonwo** [comment]
Hooray!

About autocompletion for database name.
Case is important. And quotes. It does not work in all this cases?

1."select * from MT.(autocomplete)"
2."select * from "MT".(autocomplete)"
3."select * from mt.(autocomplete)"

**Comment by Lonniebiz** [comment]
@LonwoLonwo If I capitalize "MT." auto-completion does indeed work after the dot. However, Pervasive doesn't care about capitalization of "mt." for query execution.

1."select * from MT.(autocomplete)"  **(works)**
2."select * from "MT".(autocomplete)" **(doesn't work)**
3."select * from mt.(autocomplete)" **(doesn't work)**

Back in the year 2000, when I first learn SQL, I use to capitalize keywords in my SQL. These days, I keep everything lowercase; Dbeaver automatically colorizes keywords and I no longer burden myself with the old practice of capitalizing things. Even if the table, in Pervasive, contains capital letters in its official name, Pervasive ignores capitalization in queries you submit to it. I do not capitalize anything. 

As a matter of fact, I set Dbeaver to do everything in lowercase a long time ago. See this: https://github.com/dbeaver/dbeaver/issues/839

**Comment by AKG1004** [comment]
@LonwoLonwo I am not getting any names or value under Advance parameter. I am using postgres

**Comment by LonwoLonwo** [comment]
Hello @AKG1004 

Can you provide some footage of your issue, please?

**Comment by Lonniebiz** [comment]
@LonwoLonwo 

Sure. Here's a short video:
https://youtu.be/Novu0EM7qfQ

During the video, I noticed two things:

1. The only example (listed previously) that doesn't autocomplete, is Example 3 (where I'm trying to auto-complete after a **lowercase** database/schema name).
2. Table completion (when no database/schema is specified/typed) only happens when you press `ctrl+spacebar`. I guess this because auto-completion is typically triggered by typing a dot and not dot is typed under this circumstance.

**Update:** I just now noticed that you were actually requesting for @AKG1004 to provide a video (not me). Sorry. I hope my video is still useful in showing the current behavior from a **Pervasive SQL** perspective. It seems AKG1004 will be providing a **PostgreSQL** perspective instead.

**Comment by geoartkoszalin** [comment]
You seem to be having trouble with column name autocompletion in DBeaver 23.3.0, especially when joining another table. Can you provide more details about the issue? Are there any errors or error messages you're encountering?

sample code:

SELECT 
aw.ID
--count(*)
,aw.TERYT_ULC 
,aw.ULICA 
,aw.NR_PORZADKOWY
,daw.ID_DZE
,d.kod_obr
,d.numer
/*,aw.NAZWA_WLASNA*/ 
FROM ADRESY_W aw
/*WHERE 
--aw.TERYT_ULC IS NOT NULL
aw.ULICA LIKE 'Pełki' 
GROUP BY aw.TERYT_ULC, aw.ULICA ,aw.NR_PORZADKOWY*/   
JOIN DZIALKI_ADRESY_W daw ON daw.id_ADR = aw.ID_O   
left JOIN EG_EGBIL.DZIALKI d ON d.ID_O = daw.ID_DZE
ORDER BY 3 DESC

when i'm usin alias ex aw. , daw. etc Dbeaver tell me no proposal

![porposal](https://github.com/dbeaver/dbeaver/assets/150658057/3ee7083f-d441-425e-9ba0-41be09277483)

![no proposal](https://github.com/dbeaver/dbeaver/assets/150658057/a42b4f3a-5e91-4557-9361-c8f8de2d5fd8)


Please help

**Comment by fuludag** [comment]
Hi @geoartkoszalin, did you fix this problem?

**Comment by metatron-lives** [comment]
This seems to still be a problem. 

DBeaver: Version 24.3.2.202501051917
MariaDB: Mariadb-java-client: release 3.3.2.

In MariaDB running on my local computer, tables that I create cannot be autocompleted, even when I specify schema. When I drop into the DBeaver Sample Database in SQLite, however, autocomplete with new tables works fine (as does autocomplete with the sample's already existing tables). 

MariaDB code: 
```CREATE TABLE Demonstration  
(PatientID varchar(255) NOT NULL PRIMARY KEY ,
PatientName varchar(255) NOT NULL,
PatientState varchar(255) NOT NULL,
Gender varchar(255) NOT NULL,
Visits int Null 
,Charges int Null Default 0)

INSERT INTO Demonstration 
           (PatientID
           ,PatientName
           ,PatientState
           ,Gender
           ,Visits
		   ,Charges)
     VALUES
	 ('12345','John','AL','M',3,200)
	,('12346','Jane','AK','F',1,400);

SELECT * FROM Demonstration ;
```

The select command works, and the table shows up in database finder when I look at it, but no autocomplete (including specifing schema name) works. Table names also show up in the parser as underlined erros with `"Table <name> not found"`

**Comment by CasTrooooGI** [comment]
Solución

![Image](https://github.com/user-attachments/assets/27e1f5ff-fa1f-41c0-860f-f4c2e4481126)

**Comment by CasTrooooGI** [comment]
Versión: 25.0.1 
PostgreSQL

A mi me ha funcionado cambando semantic por legacy


![Image](https://github.com/user-attachments/assets/55188f2f-1808-4932-8e87-3d2b5b30a4f5)

**Comment by MassimoFid** [comment]
Version: 25.2.0
DB: PostgreSQL

It looks like that Semantic engine performance is not appropriate when objects to be analyzed are too many.

I also solved switching to Legacy completition engine. Instead Semantic completition takes a lot of time. Looking at "Background Tasks", "Analyzing query for proposal..." struggling because my DB as a lot of tables with many partitions.

<img width="1262" height="380" alt="Image" src="https://github.com/user-attachments/assets/8b7184ab-de69-4295-97a7-71d680d6447c" />
